File: killrvideo-service-users/src/main/java/com/killrvideo/service/user/grpc/UserManagementServiceGrpcValidator.java
package com.killrvideo.service.user.grpc;

import com.killrvideo.utils.FluentValidator;
import io.grpc.stub.StreamObserver;
import killrvideo.user_management.UserManagementServiceOuterClass.CreateUserRequest;
import killrvideo.user_management.UserManagementServiceOuterClass.GetUserProfileRequest;
import killrvideo.user_management.UserManagementServiceOuterClass.VerifyCredentialsRequest;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.stereotype.Component;

import static org.apache.commons.lang3.StringUtils.isBlank;

/**
 * Validate GRPC parameters.
 *
 * @author Cedrick LUNVEN (@clunven)
 */
@Component
public class UserManagementServiceGrpcValidator {
    private static final Logger LOGGER = LoggerFactory.getLogger(UserManagementServiceGrpcValidator.class);

    /**
     * Validate create user.
     *
     * @param request        request.
     * @param streamObserver response.
     */
    public void validateGrpcRequest_createUser(CreateUserRequest request, StreamObserver<?> streamObserver) {
        FluentValidator.of("createUser", request, LOGGER, streamObserver)
                .notEmpty("user id", isBlank(request.getUserId().getValue()))
                .notEmpty("password", isBlank(request.getPassword()))
                .notEmpty("email", isBlank(request.getEmail()))
                .validate();
    }

    public void validateGrpcRequest_VerifyCredentials(VerifyCredentialsRequest request, StreamObserver<?> streamObserver) {
        FluentValidator.of("verifyCredentials", request, LOGGER, streamObserver)
                .notEmpty("email", isBlank(request.getEmail()))
                .notEmpty("password", isBlank(request.getPassword()))
                .validate();
    }

    public void validateGrpcRequest_getUserProfile(GetUserProfileRequest request, StreamObserver<?> streamObserver) {
        FluentValidator validator = FluentValidator.of("verifyCredentials", request, LOGGER, streamObserver)
                .error("cannot get more than 20 user profiles at once for get user profile request",
                        request.getUserIdsCount() > 20);
        request.getUserIdsList().forEach(uuid ->
                validator.error("provided UUID values cannot be null or blank for get user profile request",
                        uuid == null || isBlank(uuid.getValue()))
        );
        validator.validate();
    }
}

File: killrvideo-service-users/src/main/kotlin/com/killrvideo/service/user/grpc/UserManagementServiceGrpcValidator.kt
package com.killrvideo.service.user.grpc

import com.killrvideo.utils.FluentValidator
import killrvideo.user_management.UserManagementServiceOuterClass.*
import mu.KotlinLogging
import org.apache.commons.lang3.StringUtils.isBlank
import org.springframework.stereotype.Component

/**
 * Validate GRPC parameters.
 *
 * @author Cedrick LUNVEN (@clunven)
 */
@Component
open class UserManagementServiceGrpcValidator {
    private val logger = KotlinLogging.logger { }
    /**
     * Validate create user.
     *
     * @param request        request.
     */
    fun validateGrpcRequest_createUser(request: CreateUserRequest) {
        FluentValidator.of("createUser", request, logger)
            .notEmpty("user id", isBlank(request.userId.value))
            .notEmpty("password", isBlank(request.password))
            .notEmpty("email", isBlank(request.email))
            .validate()
    }

    fun validateGrpcRequest_VerifyCredentials(request: VerifyCredentialsRequest) {
        FluentValidator.of("verifyCredentials", request, logger)
            .notEmpty("email", isBlank(request.email))
            .notEmpty("password", isBlank(request.password))
            .validate()
    }

    fun validateGrpcRequest_getUserProfile(request: GetUserProfileRequest) {
        val validator: FluentValidator = FluentValidator.of("verifyCredentials", request, logger)
            .error(
                "cannot get more than 20 user profiles at once for get user profile request",
                request.userIdsCount > 20
            )
        request.userIdsList.forEach { uuid ->
            validator.error(
                "provided UUID values cannot be null or blank for get user profile request",
                uuid == null || isBlank(uuid.value)
            )
        }
        validator.validate()
    }
}
